### PR TITLE
Added Async HTML Tokenizer

### PR DESCRIPTION
**Pull request by nikhilshagri**
Design: I realized having two different parsers for sync and async was wrong, because the API for both was fundamentally the same. All I needed to do was create another Tokenizer, because `ParseNode` ( representation for nodes which are yet to be created) is used by the TreeBuilder and the Sink, and the `Tokenizer` is the 'lowermost' type concerned with these two types.

Therefore, I created one and placed it in `async_html.rs`, and also created a new Sink which deals with `ParseNode`s. I changed the methods in ServoParser to take an `async` argument too, which decides which Tokenizer will be used. The Tokenizer isn't exactly *async* for now, but this PR separates action creation from execution, which allows the async behaviour to be implemented later. Right now, all actions are executed soon after they are created.

The Sink consists of two Hashmaps, `nodes`, which contains the actual nodes, with the key being their corresponding `ParseNode`'s id, and `parse_node_data`, which contains metadata about the nodes. 

It's still a bit rough, (I can't figure out how to deal with `complete_script` and `is_mathml_annotation_xml_integration_point`, along with some other parts I wrote in a hurry), but I believe the overall design is sound. I'd like to hear what you think about it.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17037)

### CODE REVIEW

**Comment by cynicaldevil** [comment]
cc @jdm @nox

File: components/script/dom/servoparser/mod.rs [hunk 1]
```diff
@@ -91,6 +92,7 @@ pub struct ServoParser {
     aborted: Cell<bool>,
     /// https://html.spec.whatwg.org/multipage/#script-created-parser
     script_created_parser: bool,
+    async: bool,
```
Comment (nox) [inline components/script/dom/servoparser/mod.rs]: You don't need this, the provided tokeniser already lets you know whether it's async or not, please remove.

File: components/script/dom/servoparser/mod.rs [hunk 2]
```diff
@@ -54,6 +54,7 @@ use std::mem;
 use style::context::QuirksMode as ServoQuirksMode;
 
 mod html;
+mod async_html;
```
Comment (nox) [inline components/script/dom/servoparser/mod.rs]: Nit: alphabetic order.

File: components/script/dom/element.rs [hunk 3]
```diff
@@ -1364,7 +1364,7 @@ impl Element {
         // Steps 1-2.
         let context_document = document_from_node(self);
         // TODO(#11995): XML case.
-        let new_children = ServoParser::parse_html_fragment(self, markup);
+        let new_children = ServoParser::parse_html_fragment(self, markup, true);
```
Comment (nox) [inline components/script/dom/element.rs]: Why async? The only input here comes from `document.write`.

File: components/script/dom/servoparser/async_html.rs [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+        &self.inner.sink().sink().base_url
+    }
+
+    pub fn set_plaintext_state(&mut self) {
+        self.inner.set_plaintext_state();
+    }
+}
+
+#[allow(unsafe_code)]
+unsafe impl JSTraceable for HtmlTokenizer<TreeBuilder<ParseNode, Sink>> {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: This impl can be shared with the sync tokeniser.

File: components/script/dom/servoparser/async_html.rs [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+    pub context_elem: &'a ParseNode,
+    pub form_elem: Option<&'a ParseNode>,
+}
+
+#[derive(JSTraceable, Clone)]
+pub struct ParseNode {
+    id: usize,
+}
+
+enum ParserNodeOrText {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Why do you need this new enum?

File: components/script/dom/servoparser/async_html.rs [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+            script: Default::default(),
+            parse_node_data: DOMRefCell::new(HashMap::new()),
+            next_parse_node_id: Cell::new(1),
+            nodes: DOMRefCell::new(HashMap::new()),
+        }
+    }
+
+    fn new_parse_node(&self) -> ParseNode {
+        let id = self.next_parse_node_id.get();
+        // TODO: Default::default() for ParseNodeData
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: You defined that type yourself, so why a todo?

File: components/script/dom/servoparser/async_html.rs [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+    type Handle = ParseNode;
+
+    fn get_document(&mut self) -> Self::Handle {
+        ParseNode {
+            id: 0
+        }
+    }
+
+    fn get_template_contents(&mut self, target: &Self::Handle) -> Self::Handle {
+        let contents = self.new_parse_node();
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: That means every time you get the template contents, a new parse node is created. That is wrong, the same one should be returned every time.

File: components/script/dom/servoparser/async_html.rs [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+        self.enqueue(ParseOperation::GetTemplateContents(target.id, contents.id));
+        contents
+    }
+
+    fn same_node(&self, x: &Self::Handle, y: &Self::Handle) -> bool {
+        x.id == y.id
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: This doesn't work for template contents.

File: components/script/dom/servoparser/async_html.rs [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+
+    fn create_pi(&mut self, target: StrTendril, data: StrTendril) -> ParseNode {
+        let node = self.new_parse_node();
+        match self.parse_node_data.borrow_mut().get_mut(&node.id) {
+            Some(ref mut node_data) => {
+                node_data.target = Some(String::from(target));
+                node_data.data = Some(String::from(data));
+                self.enqueue(ParseOperation::CreatePI(node.id));
+            },
+            None => panic!("Element was just created using new_parse_node() and should exist!"),
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: What is this for?

File: components/script/dom/servoparser/async_html.rs [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+        };
+        node
+    }
+
+    fn has_parent_node(&self, node: &Self::Handle) -> bool {
+        match self.parse_node_data.borrow().get(&node.id) {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Use `expect` on that, instead of matching explicitly.

File: components/script/dom/servoparser/mod.rs [hunk 11]
```diff
@@ -100,16 +102,22 @@ enum LastChunkState {
...
                                       LastChunkState::NotReceived,
-                                      ParserKind::Normal);
+                                      ParserKind::Normal,
+                                      async);
         parser.parse_chunk(String::from(input));
     }
 
     // https://html.spec.whatwg.org/multipage/#parsing-html-fragments
-    pub fn parse_html_fragment(context: &Element, input: DOMString) -> FragmentParsingResult {
+    pub fn parse_html_fragment(context: &Element, input: DOMString, async: bool) -> FragmentParsingResult {
```
Comment (nox) [inline components/script/dom/servoparser/mod.rs]: Why would you parse a fragment in async mode?

File: components/script/dom/servoparser/mod.rs [hunk 12]
```diff
@@ -157,11 +171,18 @@ impl ServoParser {
         }
     }
 
-    pub fn parse_html_script_input(document: &Document, url: ServoUrl, type_: &str) {
+    pub fn parse_html_script_input(document: &Document, url: ServoUrl, type_: &str, async: bool) {
```
Comment (nox) [inline components/script/dom/servoparser/mod.rs]: Why would you parse script input in async mode?

File: components/script/dom/servoparser/mod.rs [hunk 13]
```diff
@@ -100,16 +102,22 @@ enum LastChunkState {
 }
 
 impl ServoParser {
-    pub fn parse_html_document(document: &Document, input: DOMString, url: ServoUrl) {
-        let parser = ServoParser::new(document,
-                                      Tokenizer::Html(self::html::Tokenizer::new(document, url, None)),
+    pub fn parse_html_document(document: &Document, input: DOMString, url: ServoUrl, async: bool) {
```
Comment (nox) [inline components/script/dom/servoparser/mod.rs]: No boolean argument for that please, and I think there is no reason an HTML document should be parsed in a sync way.

**Review by nox — CHANGES_REQUESTED** [review]
Initial partial review.

File: components/script/dom/servoparser/async_html.rs (hunk 7 above)
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: The name is actually misleading, the original function called the `Content` method on the target, which creates a new document fragment whenever called: https://doc.servo.org/src/script/dom/htmltemplateelement.rs.html#53

File: components/script/dom/servoparser/async_html.rs (hunk 8 above)
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: How so?

File: components/script/dom/servoparser/async_html.rs (hunk 7 above)
Comment (jdm) [inline components/script/dom/servoparser/async_html.rs]: or_init means that the node is created the first time Content is called, but cached for use by subsequent calls.

File: components/script/dom/servoparser/async_html.rs [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+
+    fn get_document(&mut self) -> Self::Handle {
+        self.document_node.clone()
+    }
+
+    fn get_template_contents(&mut self, target: &Self::Handle) -> Self::Handle {
+        let mut parse_node_data = self.parse_node_data.borrow_mut();
+        let mut data = parse_node_data.get_mut(&target.id).expect("Expected data of parse node");
+        let contents = match data.contents {
+            Some(ref node) => node.clone(),
```
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: IUIC, we can safely clone the `ParseNode` to be returned from this function. I am not able to see the problem with it, but I would like your advice on this one.

File: components/script/dom/servoparser/async_html.rs [hunk 15]
```diff
@@ -0,0 +1,546 @@
...
+        let tree_builder = &self.sink;
+        tree_builder.trace_handles(&tracer);
+        tree_builder.sink.trace(trc);
+    }
+}
+
+#[derive(JSTraceable, Clone, HeapSizeOf)]
+pub struct ParseNode {
+    id: usize,
+    qual_name: Option<QualName>,
```
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: I decided to move this field here from `ParseNodeData`, because that was the only way I could get `elem_name()` to work, given its lifetime constraints.

File: components/script/dom/servoparser/async_html.rs [hunk 16]
```diff
@@ -0,0 +1,546 @@
...
+        data.parent = None;
+        self.enqueue(ParseOperation::RemoveFromParent(target.id));
+    }
+
+    fn mark_script_already_started(&mut self, node: &Self::Handle) {
+        self.enqueue(ParseOperation::MarkScriptAlreadyStarted(node.id));
+    }
+
+    // TODO: Handle this
+    fn complete_script(&mut self, node: &Self::Handle) -> NextParserState {
```
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: I'm still not exactly sure how to handle this.

File: components/script/dom/servoparser/async_html.rs [hunk 17]
```diff
@@ -0,0 +1,546 @@
...
+            if data.parent == Some(node.id) {
+                data.parent = Some(new_parent.id);
+            }
+        }
+        self.enqueue(ParseOperation::ReparentChildren(node.id, new_parent.id));
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#html-integration-point
+    /// Specifically, the <annotation-xml> cases.
+    fn is_mathml_annotation_xml_integration_point(&self, handle: &Self::Handle) -> bool {
```
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: This one too.

**Review by nikhilshagri — COMMENTED** [review]
@nox Right! I changed most of the things you requested, and got the wikipedia main page and my github profile page (also a few pages I selected from `tests`) to run successfully at last! I'm still not sure how to implement a couple of methods from TreeSink, though.

File: components/script/dom/servoparser/async_html.rs (hunk 8 above)
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: You don't return the same index every time one call `get_template_contents`, so this method can return `false` when comparing a template contents with itself.

File: components/script/dom/servoparser/async_html.rs (hunk 17 above)
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: Fixed this one today.

File: components/script/dom/servoparser/async_html.rs [hunk 18]
```diff
@@ -0,0 +1,553 @@
...
+        let parse_node_data = self.parse_node_data.borrow();
+        let node_data = parse_node_data.get(&handle.id).expect("Element does not exist!");
+        node_data.is_integration_point
+    }
+
+    fn set_current_line(&mut self, line_number: u64) {
+        self.current_line = line_number;
```
Comment (jdm) [inline components/script/dom/servoparser/async_html.rs]: This won't work as written - the code that creates elements relies on the line number that's stored, and it will be incorrect when elements are created after the synchronous section of the tokenizer.

**Comment by cynicaldevil** [comment]
@nox I think you can review it in-depth now. I am still unsure of what to do with `complete_script`, though.

File: components/script/dom/servoparser/async_html.rs [hunk 19]
```diff
@@ -0,0 +1,561 @@
...
+                self.nodes.borrow_mut().insert(id, JS::from_ref(elem.upcast()));
+            }
+            ParseOperation::CreateComment(text, id) => {
+                let comment = Comment::new(DOMString::from(text), document);
+                assert!(self.nodes.borrow_mut().insert(id, JS::from_ref(&comment.upcast())).is_none());
+            }
+            ParseOperation::Insert(parent, reference_child, new_node) => {
+                let nodes = self.nodes.borrow();
+                let parent = nodes.get(&parent).unwrap();
+                let reference_child = reference_child.map(|n| nodes.get(&n).unwrap().deref());
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Why the explicit `deref` call?

File: components/script/dom/servoparser/async_html.rs [hunk 20]
```diff
@@ -0,0 +1,561 @@
...
+            target: None,
+            data: None,
+            contents: None,
+            is_integration_point: false,
+        }
+    }
+}
+
+enum ParseOperation {
+    GetTemplateContents(usize, usize),
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Please use named fields or document what each `usize` is. Named fields would be better of course.

File: components/script/dom/servoparser/async_html.rs [hunk 21]
```diff
@@ -0,0 +1,561 @@
...
+        };
+        let data: ParseNodeData = Default::default();
+        assert!(sink.parse_node_data.borrow_mut().insert(0, data).is_none());
+        sink.nodes.borrow_mut().insert(0, JS::from_ref(document.upcast()));
+        sink
+    }
+
+    fn new_parse_node(&self) -> ParseNode {
+        let id = self.next_parse_node_id.get();
+        let data: ParseNodeData = Default::default();
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Nit: that could be `ParseNodeData::default()`.

File: components/script/dom/servoparser/async_html.rs [hunk 22]
```diff
@@ -0,0 +1,561 @@
...
+}
+
+#[derive(JSTraceable, HeapSizeOf)]
+#[must_root]
+pub struct Sink {
+    base_url: ServoUrl,
+    document: JS<Document>,
+    current_line: u64,
+    script: MutNullableJS<HTMLScriptElement>,
+    parse_node_data: DOMRefCell<HashMap<usize, ParseNodeData>>,
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Why does this need to be in a `DOMRefCell` given all `TreeSink` methods take a `&mut self`?

File: components/script/dom/servoparser/async_html.rs [hunk 23]
```diff
@@ -0,0 +1,561 @@
...
+        let data: ParseNodeData = Default::default();
+        assert!(self.parse_node_data.borrow_mut().insert(id, data).is_none());
+        self.next_parse_node_id.set(id + 1);
+        ParseNode {
+            id: id,
+            qual_name: None,
+        }
+    }
+
+    fn enqueue(&self, op: ParseOperation) {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: I think you can remove this method or `process_operation`. :)

File: components/script/dom/servoparser/async_html.rs [hunk 24]
```diff
@@ -0,0 +1,561 @@
...
+        self.process_operation(op);
+    }
+
+    fn process_operation(&self, op: ParseOperation) {
+        let document = Root::from_ref(&**self.nodes.borrow().get(&0).expect("document node should exist!"));
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: I wonder if the data for the document node shouldn't live outside the `nodes` map.

File: components/script/dom/servoparser/async_html.rs [hunk 25]
```diff
@@ -0,0 +1,561 @@
...
+                assert!(self.nodes.borrow_mut().insert(contents, JS::from_ref(template.Content().upcast())).is_none());
+            }
+            ParseOperation::CreateElement(id, name, attrs) => {
+                let elem = Element::create(name, &*self.document,
+                                           ElementCreator::ParserCreated(self.current_line));
+                for attr in attrs {
+                    elem.set_attribute_from_parser(attr.name, DOMString::from(String::from(attr.value)), None);
+                }
+
+                self.nodes.borrow_mut().insert(id, JS::from_ref(elem.upcast()));
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Maybe make a private method on `self` named `insert_new_node`?

File: components/script/dom/servoparser/async_html.rs [hunk 26]
```diff
@@ -0,0 +1,561 @@
...
+                let parent = nodes.get(&parent).unwrap();
+                let reference_child = reference_child.map(|n| nodes.get(&n).unwrap().deref());
+
+                let node = match new_node {
+                    ParserNodeOrText::Node(ref id) => {
+                        let deref_node = nodes.get(id).unwrap().deref();
+                        NodeOrText::AppendNode(JS::from_ref(deref_node))
+                    },
+                    ParserNodeOrText::Text(ref string) => {
+                        match Tendril::from_str(string) {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Nit: you could use `expect` here.

File: components/script/dom/servoparser/async_html.rs [hunk 27]
```diff
@@ -0,0 +1,561 @@
...
+                            Ok(tendril) => NodeOrText::AppendText(tendril),
+                            Err(_) => panic!("Error while converting from string to tendril!"),
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Please file a follow-up issue on Tendril to implement `From<&str>` for `StrTendril`.

File: components/script/dom/servoparser/async_html.rs [hunk 28]
```diff
@@ -0,0 +1,561 @@
...
+            }
+            ParseOperation::AppendDoctypeToDocument(name, public_id, system_id) => {
+                let doctype = DocumentType::new(
+                    DOMString::from(name), Some(DOMString::from(public_id)),
+                    Some(DOMString::from(system_id)), document);
+
+                document.upcast::<Node>().AppendChild(doctype.upcast()).expect("Appending failed");
+            }
+            ParseOperation::AddAttrsIfMissing(target_id, attrs) => {
+                match self.nodes.borrow().get(&target_id) {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Make a private `node` method on `self` that returns a node from the `nodes` map or panics if there are none.

File: components/script/dom/servoparser/async_html.rs [hunk 29]
```diff
@@ -0,0 +1,561 @@
...
+                            parent.RemoveChild(&*target).unwrap();
+                        }
+                    }
+                    None => panic!("Node not found!")
+                }
+            }
+            ParseOperation::MarkScriptAlreadyStarted(node) => {
+                match self.nodes.borrow().get(&node) {
+                    Some(node) => {
+                        let script = node.downcast::<HTMLScriptElement>();
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Use `expect` here, this operation should be done only with `HTMLScriptElement` values anyway.

File: components/script/dom/servoparser/async_html.rs [hunk 30]
```diff
@@ -0,0 +1,561 @@
...
+                        script.map(|script| script.set_already_started(true));
+                    }
+                    None => panic!("Node not found!")
+                }
+            }
+            ParseOperation::ReparentChildren(parent, new_parent) => {
+                let nodes = self.nodes.borrow();
+                let parent = nodes.get(&parent).unwrap();
+                let new_parent = nodes.get(&new_parent).unwrap();
+                while let Some(ref child) = parent.GetFirstChild() {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Nit: `ref` useless here.

File: components/script/dom/servoparser/async_html.rs [hunk 31]
```diff
@@ -0,0 +1,561 @@
...
+            ParseOperation::AssociateWithForm(target, form) => {
+                let nodes = self.nodes.borrow();
+                let node = nodes.get(&target).unwrap();
+                let form = nodes.get(&form).unwrap();
+
+                let form = Root::downcast::<HTMLFormElement>(Root::from_ref(&**form))
+                    .expect("Owner must be a form element");
+
+                let elem = node.downcast::<Element>();
+                let control = elem.and_then(|e| e.as_maybe_form_control());
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Can this ever be called with something that isn't an `Element`?

File: components/script/dom/servoparser/async_html.rs [hunk 32]
```diff
@@ -0,0 +1,561 @@
...
+
+                if let Some(control) = control {
+                    control.set_form_owner_from_parser(&form);
+                } else {
+                    // TODO remove this code when keygen is implemented.
+                    assert!(node.NodeName() == "KEYGEN", "Unknown form-associatable element");
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Nit: `assert_eq`

File: components/script/dom/servoparser/async_html.rs [hunk 33]
```diff
@@ -0,0 +1,561 @@
...
+                }
+            }
+            ParseOperation::Pop(node) => {
+                let nodes = self.nodes.borrow();
+                let node = nodes.get(&node).unwrap();
+                let node = Root::from_ref(&**node);
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Why make a `Root` here?

File: components/script/dom/servoparser/async_html.rs [hunk 34]
```diff
@@ -0,0 +1,561 @@
...
+                vtable_for(&node).pop();
+            }
+            ParseOperation::CreatePI(node) => {
+                let parse_node_data = self.parse_node_data.borrow();
+                let data = parse_node_data.get(&node).unwrap();
+                let pi = ProcessingInstruction::new(
+                    DOMString::from(data.target.clone().unwrap()),
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Nit: instead of cloning and then wrapping, you could use `DOMString::from` with a `&str` argument.

File: components/script/dom/servoparser/async_html.rs [hunk 35]
```diff
@@ -0,0 +1,561 @@
...
+
+    fn get_template_contents(&mut self, target: &Self::Handle) -> Self::Handle {
+        let data;
+        {
+            let parse_node_data = self.parse_node_data.borrow();
+            data = parse_node_data.get(&target.id).expect("Expected data of parse node").contents.clone();
+        }
+        if data.is_some() {
+            return data.clone().unwrap();
+        }
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: I don't understand this code and the weird scoping, can you explain it to me?

File: components/script/dom/servoparser/async_html.rs [hunk 36]
```diff
@@ -0,0 +1,561 @@
...
+
+        let node = self.new_parse_node();
+        let mut parse_node_data = self.parse_node_data.borrow_mut();
+        let mut data = parse_node_data.get_mut(&target.id).expect("Expected data of parse node");
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: As with the method `node` I suggested earlier, you probably want a `node_mut` method here.

File: components/script/dom/servoparser/async_html.rs [hunk 37]
```diff
@@ -0,0 +1,561 @@
...
+        node
+    }
+
+    fn same_node(&self, x: &Self::Handle, y: &Self::Handle) -> bool {
+        x.id == y.id
+    }
+
+    fn elem_name<'a>(&self, target: &'a Self::Handle) -> ExpandedName<'a> {
+        match target.qual_name {
+            Some(ref qual_name) => qual_name.expanded(),
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Nit: you can do `target.qual_name.as_ref().expect(...).expanded()` here.

File: components/script/dom/servoparser/async_html.rs [hunk 38]
```diff
@@ -0,0 +1,561 @@
...
+            None => panic!("Expected qual name of node!"),
+        }
+    }
+
+    fn same_tree(&self, x: &Self::Handle, y: &Self::Handle) -> bool {
+        let nodes = self.nodes.borrow();
+        let (id_x, id_y) = (x.id, y.id);
+
+        let x = nodes.get(&id_x).expect("Node not found");
+        // Update x's parent if no longer in tree
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Can you explain why this is needed?

File: components/script/dom/servoparser/async_html.rs [hunk 39]
```diff
@@ -0,0 +1,561 @@
...
+    fn create_element(&mut self, name: QualName, attrs: Vec<Attribute>, _flags: ElementFlags)
+        -> Self::Handle {
+        let mut node = self.new_parse_node();
+        node.qual_name = Some(name.clone());
+
+        let mut parse_node_data = self.parse_node_data.borrow_mut();
+        let mut node_data = parse_node_data.get_mut(&node.id).expect("Element does not exist!");
+        node_data.is_integration_point = attrs.iter()
+        .any(|attr| {
+            let attr_value = &String::from(attr.value.clone());
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Why is this needed?

File: components/script/dom/servoparser/async_html.rs [hunk 40]
```diff
@@ -0,0 +1,561 @@
...
+            (attr_value.eq_ignore_ascii_case("text/html") ||
+            attr_value.eq_ignore_ascii_case("application/xhtml+xml"))
+        });
+        self.enqueue(ParseOperation::CreateElement(node.id, name, attrs));
+        node
+    }
+
+    fn create_comment(&mut self, text: StrTendril) -> Self::Handle {
+        let node = self.new_parse_node();
+        self.enqueue(ParseOperation::CreateComment(String::from(text), node.id));
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Why aren't we just storing the `StrTendril` value here?

File: components/script/dom/servoparser/async_html.rs [hunk 41]
```diff
@@ -0,0 +1,561 @@
...
+    fn parse_error(&mut self, msg: Cow<'static, str>) {
+        debug!("Parse error: {}", msg);
+    }
+
+    fn set_quirks_mode(&mut self, mode: QuirksMode) {
+        let mode = match mode {
+            QuirksMode::Quirks => ServoQuirksMode::Quirks,
+            QuirksMode::LimitedQuirks => ServoQuirksMode::LimitedQuirks,
+            QuirksMode::NoQuirks => ServoQuirksMode::NoQuirks,
+        };
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Why do we have 2 quirks mode enum again?

File: components/script/dom/servoparser/async_html.rs [hunk 42]
```diff
@@ -0,0 +1,561 @@
...
+        data.parent = None;
+        self.enqueue(ParseOperation::RemoveFromParent(target.id));
+    }
+
+    fn mark_script_already_started(&mut self, node: &Self::Handle) {
+        self.enqueue(ParseOperation::MarkScriptAlreadyStarted(node.id));
+    }
+
+    fn complete_script(&mut self, _: &Self::Handle) -> NextParserState {
+        panic!("complete_script should not be called here!");
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Please file a follow-up about killing this altogether.

File: components/script/dom/servoparser/async_html.rs (hunk 18 above)
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: This stills stands, right @cynicaldevil?

**Review by nox — CHANGES_REQUESTED** [review]
You are doing great!

File: components/script/dom/servoparser/async_html.rs (hunk 32 above)
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: This piece of code was not written by me :smile: .This exists in the impl for the sync Tokenizer too, I copied it from there.

File: components/script/dom/servoparser/async_html.rs (hunk 18 above)
Comment (jdm) [inline components/script/dom/servoparser/async_html.rs]: I misread the changes. Currently this works fine, but when we stop calling process_operation immediately after creating parse nodes it will no longer work.

File: components/script/dom/servoparser/async_html.rs (hunk 22 above)
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: This is a tricky one, `parse_node_data` needs to be `DOMRefCell` because I want to use it mutably in the `same_tree` method, where I update the tree-related info based on the data I get from the updated `Node`s (`same tree` borrows self immutably).

The `nodes` field needs it because it is used in `process_operation`, which takes &self. If `process_operation` begins taking `&mut self`, it would cause problems with the TreeSink methods, as many methods already immutably borrow `parse_node_data` there, when they call `process_operation`. Therefore, the `process_operation` must borrow immutably too, to satisfy the borrow checker.

However, I might be able to make `nodes` a normal Hashmap in the next PR, because it will exist in main thread altogether, and therefore there will be no conflicts with `Sink` then.

File: components/script/dom/servoparser/async_html.rs (hunk 39 above)
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: I found that I could compute this value once and store it when I'm creating a new element (the value's needed in `is_mathml_annotation_xml_integration_point`). But now that I think about it, this could be a problem if the element's attributes related to this value are changed at any point of time *after* the element's created. Do you think I should change it to be synchronous ? (i.e, get the actual `Node` from the nodes map and compute the value from it).

File: components/script/dom/servoparser/async_html.rs (hunk 34 above)
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: :bowtie:
giphy

File: components/script/dom/servoparser/async_html.rs [hunk 43]
```diff
@@ -0,0 +1,493 @@
...
+
+    fn create_comment(&mut self, text: StrTendril) -> Self::Handle {
+        let node = self.new_parse_node();
+        self.process_operation(ParseOperation::CreateComment(String::from(text), node.id));
+        node
+    }
+
+    fn create_pi(&mut self, target: StrTendril, data: StrTendril) -> ParseNode {
+        let node = self.new_parse_node();
+        {
```
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: Quite a few of these scopes appear in the TreeSink methods. It's needed because `self` needs to be borrowed for `parse_node_data` operations, and then in the end needs to be borrowed *mutably* for process_operation.

File: components/script/dom/servoparser/async_html.rs (hunk 4 above)
Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: Forgot about this one, trying it now.

Comment (cynicaldevil) [inline components/script/dom/servoparser/async_html.rs]: This is not possible, I tried to make the outer impl generic, but it seems that the inner impl (the one inside trace) cannot pick up the generic type.

**Comment by nox** [comment]
@bors-servo try

This is looking good.

@jdm Do we want to make the use of the async tokenizer settable through some config?

**Comment by jdm** [comment]
A preference probably makes the most sense.

**Comment by cynicaldevil** [comment]
How should I implement it?

**Comment by jdm** [comment]
Probably check the pref in parse_html_document and disregard Async if the preference is disabled.

**Comment by cynicaldevil** [comment]
Added preference.

**Review by nox — APPROVED** [review]
